File: packages/js/src/stream.ts
import { KovaTTSProtocolError } from "./errors.js";
import { decodeBase64ToBytes } from "./audio.js";
import type { StreamEvent } from "./types.js";

export function parseStreamEvent(value: unknown): StreamEvent {
  if (!value || typeof value !== "object") {
    throw new KovaTTSProtocolError("Stream event must be an object");
  }

  const event = value as Record<string, unknown>;
  if (event.type === "audio" && typeof event.audio_chunk === "string") {
    return {
      type: "audio",
      audio: decodeBase64ToBytes(event.audio_chunk),
    };
  }

  if (
    event.type === "timestamps" &&
    Array.isArray(event.words) &&
    Array.isArray(event.start_seconds) &&
    Array.isArray(event.end_seconds)
  ) {
    return {
      type: "timestamps",
      words: event.words as string[],
      start_seconds: event.start_seconds as number[],
      end_seconds: event.end_seconds as number[],
    };
  }

  throw new KovaTTSProtocolError(`Unknown stream event shape: ${JSON.stringify(value)}`);
}

export async function* parseEventStream(
  stream: ReadableStream<Uint8Array>,
): AsyncGenerator<StreamEvent> {
  const reader = stream.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (true) {
      const { value, done } = await reader.read();
      if (done) {
        break;
      }
      buffer += decoder.decode(value, { stream: true });
      const records = buffer.split(/\r?\n\r?\n/);
      buffer = records.pop() ?? "";
      for (const record of records) {
        const event = parseSseRecord(record);
        if (event) {
          yield event;
        }
      }
    }

    buffer += decoder.decode();
    const finalEvent = parseSseRecord(buffer);
    if (finalEvent) {
      yield finalEvent;
    }
  } finally {
    reader.releaseLock();
  }
}

export function parseSseRecord(record: string): StreamEvent | null {
  const data = record
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line.startsWith("data:"))
    .map((line) => line.slice(5).trim())
    .join("\n");

  if (!data) {
    return null;
  }

  return parseStreamEvent(JSON.parse(data));
}

File: packages/python/src/kova_tts/stream.py
from __future__ import annotations

import json
from collections.abc import AsyncIterator

from .errors import KovaTTSProtocolError
from .types import StreamAudioEvent, StreamEvent, StreamTimestampsEvent


def parse_stream_event(value: object) -> StreamEvent:
    if not isinstance(value, dict):
        raise KovaTTSProtocolError("Stream event must be an object")

    if value.get("type") == "audio" and isinstance(value.get("audio_chunk"), str):
        return StreamAudioEvent(type="audio", audio_chunk=value["audio_chunk"])

    if value.get("type") == "timestamps":
        words = value.get("words")
        start_seconds = value.get("start_seconds")
        end_seconds = value.get("end_seconds")
        if isinstance(words, list) and isinstance(start_seconds, list) and isinstance(end_seconds, list):
            return StreamTimestampsEvent(
                type="timestamps",
                words=words,
                start_seconds=start_seconds,
                end_seconds=end_seconds,
            )

    raise KovaTTSProtocolError(f"Unknown stream event shape: {value!r}")


def parse_sse_record(record: str) -> StreamEvent | None:
    data = "\n".join(
        line.strip()[5:].strip()
        for line in record.splitlines()
        if line.strip().startswith("data:")
    )
    if not data:
        return None
    return parse_stream_event(json.loads(data))


async def parse_event_stream(lines: AsyncIterator[str]) -> AsyncIterator[StreamEvent]:
    record_lines: list[str] = []
    async for line in lines:
        if line.strip():
            record_lines.append(line)
            continue
        event = parse_sse_record("\n".join(record_lines))
        record_lines = []
        if event is not None:
            yield event

    if record_lines:
        event = parse_sse_record("\n".join(record_lines))
        if event is not None:
            yield event
